# *A. halleri* Bayesian Causal Network Analysis 

This analysis uses the [causalnex](https://causalnex.readthedocs.io/) python package. This is an explanation of code that has only been tested in a `PyTorch` Docker container image on a server with research grade GPU's and >250 cores.

## Import Python Libraries

In [ ]:
# import pandas and numpy
import pandas as pd
import numpy as np

# setup label encoder to transform non-numeric data into numeric data
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

# silence warnings
import warnings
warnings.filterwarnings("ignore")

# import StructureModel 
from causalnex.structure import StructureModel
sm = StructureModel()


## *A. halleri Data Input and Formatting*

### Read in phenotype and environment data

These data are not normalized or scaled in anyway.

In [ ]:
data = pd.read_csv('~/a_halleri/data/Ahalleri_transplant_exp_F.txt', delimiter='\t')
data.head()

sample origin_pop  ... F_pollen_deaddegen_perc F_pollen_diameter
0  M_PL22_101     M_PL22  ...                0.607903         15.540000
1  M_PL22_101     M_PL22  ...                1.818182         16.653333
2  M_PL22_101     M_PL22  ...                1.279707         18.220000
3  M_PL22_101     M_PL22  ...                1.141553         17.278333
4  M_PL22_102     M_PL22  ...                0.722543         16.650000

[5 rows x 82 columns]

### Drop features from Model

1. Drop `code` from the model, this is internal to the researchers who collected the data.
2. Drop `comparison_type` and `comparison_pop`, by one-hot encoding `comparison_3levels` the new features account for the dropped features in their comparison.
3. Use `data` to begin encoding categorical variables.

In [ ]:
drop_col = ['code', 'comparison_type', 'comparison_pop']
data = data.drop(columns=drop_col)
data.head(5)

sample origin_pop  ... F_pollen_deaddegen_perc F_pollen_diameter
0  M_PL22_101     M_PL22  ...                0.607903         15.540000
1  M_PL22_101     M_PL22  ...                1.818182         16.653333
2  M_PL22_101     M_PL22  ...                1.279707         18.220000
3  M_PL22_101     M_PL22  ...                1.141553         17.278333
4  M_PL22_102     M_PL22  ...                0.722543         16.650000

[5 rows x 79 columns]

### Transform Categorical Variables to Numeric Ones

1. Copy the `pandas` dataframe `data` to `struct_data` for `NOTEARS` preprocessing.
2. Remove `NA` values
3. Rename `sample` to `genotype` to not conflict with the python function `sample()`.
4. Encode categoricals with `>3` categories as dummy variables. 
  * Those are: `genotype`, `origin_pop`, `treatment_pop`, and `comparison_3levels`
  * The output of this encoding is stored in the dataframe `dum_df` 
5. Use the `labelEncoder()` from `sklearn.preprocessing` to convert all two category variables to binary. 
  * `M` is encoded as a `0` and `NM` is encoded as a `1`
6. 

In [ ]:
from sklearn.preprocessing import LabelEncoder
struct_data = data.copy()

# drop na rows 
struct_data = struct_data.dropna()
# change sample to genotype to not interfere with code by invoking sample()
struct_data = struct_data.rename(columns={"sample": "genotype"})

#encode non-binary categorical variables as dummy vars
dum_df = pd.get_dummies(struct_data, columns=['genotype', 'origin_pop',
                                                  'treatment_pop',
                                                  'comparison_3levels'])
# encode binary categorical variables as 0's or 1's
non_numeric_columns = list(dum_df.select_dtypes(exclude=[np.number]).columns)
le = LabelEncoder()
for col in non_numeric_columns:
  dum_df[col] = le.fit_transform(dum_df[col])

dum_df.head(5)

origin_type  ...  comparison_3levels_sympatric
0            0  ...                             1
1            0  ...                             0
2            0  ...                             0
6            0  ...                             0
7            0  ...                             0

[5 rows x 125 columns]

## Structure Learning with `NOTEARS`

The function `causalnex.structure.notears.from_pandas()` has the following arguments:

```python
X (DataFrame) – input data.
max_iter (int) – max number of dual ascent steps during optimisation.
h_tol (float) – exit if h(W) < h_tol (as opposed to strict definition of 0).
w_threshold (float) – fixed threshold for absolute edge weights.
tabu_edges (Optional[List[Tuple[str, str]]]) – list of edges(from, to) not to be included in the graph.
tabu_parent_nodes (Optional[List[str]]) – list of nodes banned from being a parent of any other nodes.
tabu_child_nodes (Optional[List[str]]) – list of nodes banned from being a child of any other nodes.
```

The `NOTEARS` algorithm has the following references:

```latex
@inproceedings{zheng2018dags,
    author = {Zheng, Xun and Aragam, Bryon and Ravikumar, Pradeep and Xing, Eric P.},
    booktitle = {Advances in Neural Information Processing Systems},
    title = {{DAGs with NO TEARS: Continuous Optimization for Structure Learning}},
    year = {2018}
}

@inproceedings{zheng2020learning,
    author = {Zheng, Xun and Dan, Chen and Aragam, Bryon and Ravikumar, Pradeep and Xing, Eric P.},
    booktitle = {International Conference on Artificial Intelligence and Statistics},
    title = {{Learning sparse nonparametric DAGs}},
    year = {2020}
}
```

### Using `NOTEARS`: Generate Directed Acyclic Graph (DAG) of *A halleri* Data 

`max_iter`: maximum number of iterations to run NOTEARS
`w_threshold`: sets absolute edge weight


In [ ]:
from causalnex.structure.notears import from_pandas
sm = from_pandas(X=dum_df, max_iter=1000, w_threshold=0.8)

### Visualize Graph Structure

Creates DAG with `plot_structure` function from CausalNex.

In [ ]:
from causalnex.plots import plot_structure
plot = plot_structure(sm)
plot.draw("sm_plot.png")

# Current Progress

As of Apr 06 2021, the learning takes ~4 hours with 125 features and 250 cores on the CyVerse Developer Server. The graph generation throws an error message:

```bash
Exception ignored in: <function AGraph.__del__ at 0x7f49dabff430>
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/pygraphviz/agraph.py", line 283, in __del__
  File "/opt/conda/lib/python3.8/site-packages/pygraphviz/agraph.py", line 1000, in _close_handle
TypeError: 'NoneType' object is not callable
```

But the graph still output as `sm_plot.png`

# Potential Next Steps:

- [ ] Fit the Conditional Distribution of the Bayesian Network
  - Requires `train/test` split with `sklearn`
- [ ] Assess Model Quality
- [ ] Query the Marginals
